File: 3-TwoIntegerSum/solution.java
import java.util.HashMap;

class Solution {
    public int[] twoSum(int[] nums, int target) {
        // Criamos um HashMap para armazenar o número e seu respectivo índice
        HashMap<Integer, Integer> seen = new HashMap<>();
        
        // Iteramos sobre o array de números
        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i]; // Calculamos o complemento
            
            // Verificamos se o complemento já existe no HashMap
            if (seen.containsKey(complement)) {
                // Se o complemento existir, retornamos o índice do complemento e o índice atual
                return new int[] { seen.get(complement), i };
            }
            
            // Armazenamos o número atual e seu índice no HashMap
            seen.put(nums[i], i);
        }
        
        // Caso não encontre a soma, retornamos um array vazio (apesar de o problema garantir que sempre haverá uma solução)
        return new int[] {};
    }
}

File: 3-TwoIntegerSum/solution.go
package main

import "fmt"

func twoSum(nums []int, target int) []int {
    // Criamos um mapa para armazenar o número e seu respectivo índice
    seen := make(map[int]int)
    
    // Iteramos sobre o slice de números
    for i, num := range nums {
        complement := target - num // Calculamos o complemento
        
        // Verificamos se o complemento já existe no mapa
        if index, found := seen[complement]; found {
            // Se o complemento foi encontrado, retornamos os índices
            return []int{index, i}
        }
        
        // Caso contrário, armazenamos o número atual e seu índice no mapa
        seen[num] = i
    }
    
    // Caso não encontre, retornamos um slice vazio (apesar de o problema garantir que sempre há uma solução)
    return []int{}
}

func main() {
    nums := []int{2, 7, 11, 15}
    target := 9
    result := twoSum(nums, target)
    fmt.Println(result)  // Deve imprimir: [0 1]
}
